### ISSUE TITLE
Build breaks on Node v12

### ISSUE THREAD

**Issue by prarabdhb** [pmndrs/drei#397]
- `three` version: `0.128.0`
- `@react-three/fiber` version: `6.1.4`
- `@react-three/drei` version: `5.0.0`
- `node` version: `12.14.0`
- `npm` (or `yarn`) version: `7.7.6`

### Problem description:

After upgrading to v5, my project has stopped building. Has the build process for this library changed?

### Relevant code:

```
[91mFailed to compile.

[0m[91m./node_modules/@react-three/drei/web/Html.js 109:61
Module parse failed: Unexpected token (109:61)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|   const transformOuterRef = React.useRef(null);
|   const transformInnerRef = React.useRef(null);
>   const target = (portal == null ? void 0 : portal.current) ?? gl.domElement.parentNode;
|   React.useEffect(() => {
|     if (group.current) {
```

### Suggested solution:

I think due to no support of optional chaining and nullish Coalescing operators on node v12, the build breaks. Maybe the build process has changed and these operators are not being transpiled?

**Comment by joshuaellis** [comment]
I did change the build process, removed deps etc. But I specifically checked on Babel, but maybe I misunderstood what it means when it says 

> NOTE: This plugin is included in @babel/preset-env

See [here](https://babeljs.io/docs/en/babel-plugin-proposal-nullish-coalescing-operator). Do you still have to include it in the plugins section? Is the issue resolved if you include such plugin in your `babel` config?

**Comment by gtolarc** [comment]
Same problem here in CRA

**Comment by joshuaellis** [comment]
> Same problem here in CRA

not helpful

**Comment by joshuaellis** [comment]
@prarabdhb Can you provide a repo please?

**Comment by Honga1** [comment]
I'm having the same issues. I found this:
https://stackoverflow.com/questions/63251072/babel-preset-env-version-7-not-transforming-the-null-coalescing-operator

And:
https://github.com/tc39/proposal-nullish-coalescing

So I was able to make it output the polyfill by setting the rollup config to
```js
  plugins: [
    ['@babel/transform-runtime', { regenerator: false, useESModules }],
    '@babel/plugin-proposal-nullish-coalescing-operator',
  ],
```

Or alternatively:
```js
  presets: [
    [
      '@babel/preset-env',
      {
        loose: true,
        modules: false,
        targets,
        include: ['proposal-nullish-coalescing-operator', 'proposal-optional-chaining'],
      },
    ],
    '@babel/preset-react',
    '@babel/preset-typescript',
  ],
```

Adds the polyfill. But alas, including the built project breaks with complaints about:
![image](https://user-images.githubusercontent.com/33930602/118846660-3e6bf100-b8cd-11eb-9018-6591f4585fb4.png)

I'd say my issue in now unrelated to the OP. But I hope this helps.

**Comment by mohith01** [comment]
> I'm having the same issues. I found this:
> https://stackoverflow.com/questions/63251072/babel-preset-env-version-7-not-transforming-the-null-coalescing-operator
> 
> Environment:
> CRA 4.0.3
> npm 7.13.0
> node v12.16.1
> 
> Occurs by:
> Including 'Html' component.
> 
> And:
> https://github.com/tc39/proposal-nullish-coalescing
> 
> So I was able to make it output the polyfill by setting the rollup config to
> 
> ```js
>   plugins: [
>     ['@babel/transform-runtime', { regenerator: false, useESModules }],
>     '@babel/plugin-proposal-nullish-coalescing-operator',
>   ],
> ```
> 
> Or alternatively by following this https://babeljs.io/docs/en/babel-preset-env#include :
> 
> ```js
>   presets: [
>     [
>       '@babel/preset-env',
>       {
>         loose: true,
>         modules: false,
>         targets,
>         include: ['proposal-nullish-coalescing-operator', 'proposal-optional-chaining'],
>       },
>     ],
>     '@babel/preset-react',
>     '@babel/preset-typescript',
>   ],
> ```
> 
> Adds the polyfill. But alas, including the built project breaks with complaints about:
> ![image](https://user-images.githubusercontent.com/33930602/118846660-3e6bf100-b8cd-11eb-9018-6591f4585fb4.png)
> 
> I'd say my issue in now unrelated to the OP. But I hope this helps.
> 
> Edit - Repro:
> 
> ```
> npx create-react-app repro --template typescript
> cd repro
> npm i
> npm i @react-three/drei @react-three/fiber
> ```
> 
> ```tsx
> import { Html } from "@react-three/drei";
> import { Canvas } from "@react-three/fiber";
> import React from "react";
> import "./App.css";
> 
> function App() {
>   return (
>     <div className="App">
>       <Canvas>
>         <Html>Dog</Html>
>       </Canvas>
>     </div>
>   );
> }
> 
> export default App;
> ```
> 
> ```
> $ npm start
> Failed to compile.
> 
> ./node_modules/@react-three/drei/core/softShadows.js 11:40
> Module parse failed: Unexpected token (11:40)
> File was processed with these loaders:
>  * ./node_modules/react-scripts/node_modules/babel-loader/lib/index.js
> You may need an additional loader to handle the result of these loaders.
> |   rings = 11
> | } = {}) => `#define LIGHT_WORLD_SIZE ${size}
> > #define LIGHT_FRUSTUM_WIDTH ${frustrum ?? frustum}
> | #define LIGHT_SIZE_UV (LIGHT_WORLD_SIZE / LIGHT_FRUSTUM_WIDTH)
> | #define NEAR_PLANE ${near}
> ```

I got this error when i use Orbitcontrols
```py
import "./styles.css"
import { Canvas } from "@react-three/fiber";
import {OrbitControls } from "@react-three/drei";

function Box(){
  return (
    <mesh>
      <boxBufferGeometry attach="geometry" />
      <meshLambertMaterial attach="material" color="hotpink" />
    </mesh>
  )
}

export default function App() {
  return (
  <Canvas>
    <OrbitControls/>
    <Box />
  </Canvas> 
  );
}
```

Error
```
Failed to compile.

./node_modules/@react-three/drei/core/softShadows.js 11:40
Module parse failed: Unexpected token (11:40)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   rings = 11
| } = {}) => `#define LIGHT_WORLD_SIZE ${size}
> #define LIGHT_FRUSTUM_WIDTH ${frustrum ?? frustum}
| #define LIGHT_SIZE_UV (LIGHT_WORLD_SIZE / LIGHT_FRUSTUM_WIDTH)
| #define NEAR_PLANE ${near}
```

**Comment by Epiczzor** [comment]
Hey i was facing the above issue, it seems to be happening in all versions after 5, i have rolled back to version ^4.1.2 and it is working fine with the rest of the packages

**Comment by d3fkon** [comment]
Updating the babel config works too 👍

**Comment by joshuaellis** [comment]
I don't think we should be transpiling our code for publishing. It makes it harder to debug for everyone involved. I think what you should be doing is asking `react-scripts` to include `proposal-nullish-coalescing-operator` in their `babel` config. Nextjs do it and i've never had an issue.

**Comment by prarabdhb** [comment]
I don't think the issue is just the `nullish-coalescing-operator` plugin. Even if we include that in the babel config, the issue pointed out by @Honga1 exists. As it can be seen that the JS file includes code from TS.

**Comment by joshuaellis** [comment]
@prarabdhb does it? Where?

**Comment by lowellwashere2020** [comment]
Saw mine was closed and moved here to where its happening to others. I'm new to all of this. What is the best way to resolve this issue?

**Comment by joshuaellis** [comment]
See Honga's comment about Babel plugins

**Comment by prarabdhb** [comment]
Hey. Wasn't able to work on a repro. Will probably look into it soon. Although, I saw what you said in your comment regarding NextJS using the plugin. I'm actually working with NextJS itself, have `next/babel` as my babel plugin. So I'm not sure why is this not working. 

Also, apologies for my earlier comment about TS code being included. Got confused there for a bit.

**Comment by prarabdhb** [comment]
@joshuaellis Here is a repro
https://github.com/prarabdhb/next-drei-repro

Clone
`yarn`
`yarn build`

The build fails

**Comment by lowellwashere2020** [comment]
> See Honga's comment about Babel plugins

I'll use this with his comments to see if I cannot figure it out. Thus far following what he did hasn't worked. but I'm probably doing it wrong :/
https://babeljs.io/docs/en/plugins/

**Comment by Honga1** [comment]
> > See Honga's comment about Babel plugins
> 
> I'll use this with his comments to see if I cannot figure it out. Thus far following what he did hasn't worked. but I'm probably doing it wrong :/
> https://babeljs.io/docs/en/plugins/

Just to clarify my fix was on this repo. Not on your own. So I cloned it, built it, and included the local npm package in my projects package.json, by using a file path dependency.

**Comment by krimpenrik** [comment]
If am having this issue as well.
```

  "dependencies": {
    "@react-three/drei": "5.0.0",
    "@react-three/fiber": "^6.1.5",
    "@svgr/webpack": "^5.5.0",
    "autoprefixer": "^10.2.5",
    "file-loader": "^6.2.0",
    "firebase": "^8.4.2",
    "firebase-admin": "^9.6.0",
    "next": "10.x",
    "postcss": "^8.2.12",
    "react": "17.x",
    "react-dom": "17.x",

    "three": "^0.128.0"
  }
```

**Comment by joshuaellis** [comment]
Can you all try `drei@5.1.5-beta.1` please. If it worked, a thumbs up will be fine, if it didn't please only comment if no one else has.

**Comment by joshuaellis** [comment]
And @prarabdhb if it solves your issue, can you please close it.

**Comment by joshuaellis** [comment]
this has been released as `5.1.6`
